util/tracing/snax_event_generator.py

```diff
@@ -0,0 +1,142 @@
+from dataclasses import dataclass, field
+
+from compiler.dialects.accfg import AcceleratorOp
+from util.tracing.annotation import EventGenerator
+from util.tracing.event import DurationEvent
+from util.tracing.state import CSRInstruction, TraceState
+
+
+@dataclass
+class _CurrentEvent:
+    start: int
+
+
+@dataclass
+class SettingUp(_CurrentEvent):
+    number_setups: int = field(default=1)
+    number_of_launches: int = field(default=0)
+    """
+    Since we only end this region when we see the *second* write to launch, we need to keep track of the number
+    of launches
+    """
+
+
+@dataclass
+class Launched(_CurrentEvent):
+    number_of_zero_writes: int = field(default=0)
+    """
+    How many zero writes to the launch address have happened
+    """
+    number_setups: int = field(default=0)
+
+
+@dataclass
+class Stalled(_CurrentEvent):
+    pass
+
+
+class SNAXAcceleratorEventGenerator(EventGenerator):
+    """
+    Works for the current version of SNAX.
+
+    We emit a "pre-launch" event, that:
+        - starts at the first CSR write to a config register
+        - ends after all launch fields have been written to
+        - records the number of setup instructions executed
+    We emit a "launched but not waiting" event, that:
+        - starts immediately after the "pre-launch" event
+        - ends on the second "write 0 to launch addr"
+    We emit a "waiting" event, that:
+        - starts on the first read from the barrier address
+        - ends when that read return 1
+
+
+    """
+
+    state: _CurrentEvent | None
+    acc: AcceleratorOp
+    fields: set[int]
+    launch_fields: set[int]
+    barrier_addr: int
+
+    def __init__(self, acc_op: AcceleratorOp):
+        super().__init__()
+        self.state = None
+        self.acc = acc_op
+        self.fields = {val.value.data for val in self.acc.fields.data.values()}
+        self.launch_fields = {
+            val.value.data for val in self.acc.launch_fields.data.values()
+        }
+        self.barrier_addr = self.acc.barrier.value.data
+
+    def cycle(self, state: TraceState) -> list[DurationEvent]:
+        ins = state.instruction
+        if not isinstance(ins, CSRInstruction):
+            return []
+        events: list[DurationEvent] = []
+
+        # if state is None, check if we are setting up, if so, switch state to "setting up"
+        if self.state is None:
+            if (ins.csr in self.fields) or (ins.csr in self.launch_fields):
```

Comment:
This was a bit wrong before :smile: :
```
            if ins.csr in self.fields or self.launch_fields):
```
Which is almost always true :sweat_smile: